src/hotspot/cpu/aarch64/templateTable_aarch64.cpp

```diff
@@ -3242,54 +3292,24 @@ void TemplateTable::fast_xaccess(TosState state)
 //-----------------------------------------------------------------------------
 // Calls
 
-void TemplateTable::prepare_invoke(int byte_no,
-                                   Register method, // linked method (or i-klass)
-                                   Register index,  // itable index, MethodType, etc.
-                                   Register recv,   // if caller wants to see it
+void TemplateTable::prepare_invoke(Register recv,   // if caller wants to see it
                                    Register flags   // if caller wants to test it
                                    ) {
-  // determine flags
+
+  const Register cache = r2;
   Bytecodes::Code code = bytecode();
-  const bool is_invokeinterface  = code == Bytecodes::_invokeinterface;
-  const bool is_invokedynamic    = code == Bytecodes::_invokedynamic;
-  const bool is_invokehandle     = code == Bytecodes::_invokehandle;
-  const bool is_invokevirtual    = code == Bytecodes::_invokevirtual;
-  const bool is_invokespecial    = code == Bytecodes::_invokespecial;
-  const bool load_receiver       = (recv  != noreg);
-  const bool save_flags          = (flags != noreg);
-  assert(load_receiver == (code != Bytecodes::_invokestatic && code != Bytecodes::_invokedynamic), "");
-  assert(save_flags    == (is_invokeinterface || is_invokevirtual), "need flags for vfinal");
-  assert(flags == noreg || flags == r3, "");
-  assert(recv  == noreg || recv  == r2, "");
-
-  // setup registers & access constant pool cache
-  if (recv  == noreg)  recv  = r2;
-  if (flags == noreg)  flags = r3;
-  assert_different_registers(method, index, recv, flags);
+  const bool load_receiver       = (code != Bytecodes::_invokestatic) && (code != Bytecodes::_invokedynamic);
+  assert_different_registers(recv, flags);
 
   // save 'interpreter return address'
-  __ save_bcp();
+  __ save_bcp(); // probably don't need this
```

Comment:
I'm not sure why you added this comment but we do need this call to `save_bcp()`. It records in the interpreter frame the 'return address' needed for re-entry into the interpreter. We cannot rely on the value cached in register rbcp remaining constant across the invoke that is about to happen.